net/libfab/libfab.c

```diff
@@ -0,0 +1,2829 @@
+	m0_nep_tlink_init_at_tail(net, &tm->ftm_ntm->ntm_end_points);
+	net->nep_addr = (const char *)(&ep->fep_name.fen_str_addr);
+	m0_ref_init(&ep->fep_nep.nep_ref, 1, &libfab_ep_release);
+	
+	return M0_RC(rc);
+}
+
+/**
+ * Used to create a passive endpoint which will
+ * listen for incoming connection requests. (Server)
+ */
+static int libfab_passive_ep_create(struct m0_fab__ep *ep, 
+				    struct m0_fab__tm *tm)
+{
+	struct m0_fab__passive_ep *pep;
+	struct fi_info            *hints;
+	struct fi_info            *fi;
+	int                        i;
+	int                        rc;
+	char                      *port = NULL;
+
+	M0_ENTRY("ep=%s addr=%s port=%s", (char*)ep->fep_name.fen_str_addr,
+		 (char*)ep->fep_name.fen_addr, (char*)ep->fep_name.fen_port);
+
+	M0_ALLOC_PTR(ep->fep_listen);
+	if (ep->fep_listen == NULL)
+		return M0_RC(-ENOMEM);
+	M0_ALLOC_PTR(ep->fep_listen->pep_aep);
+	if (ep->fep_listen->pep_aep == NULL)
+		return M0_RC(-ENOMEM);
+
+	pep = ep->fep_listen;
+	ep->fep_listen->pep_aep->aep_rxep = NULL;
+	ep->fep_listen->pep_aep->aep_txep = NULL;
+
+	if (strlen(ep->fep_name.fen_port) != 0)
+		port = ep->fep_name.fen_port;
+
+	hints = fi_allocinfo();
+	if (hints == NULL)
+		return M0_ERR(-ENOMEM);
+
+	hints->ep_attr->type = FI_EP_MSG;
+	hints->caps = FI_MSG | FI_RMA;
+	hints->domain_attr->cq_data_size = 4;
+
+	for (i = 0; i < ARRAY_SIZE(providers); i++) {
+		hints->fabric_attr->prov_name = providers[i];
+		rc = fi_getinfo(LIBFAB_VERSION, NULL, port, FI_SOURCE, hints,
+				&fi);
+		if (rc == FI_SUCCESS)
+			break;
+	}
+
+	M0_ASSERT(i < ARRAY_SIZE(providers));
+	hints->fabric_attr->prov_name = NULL;
+	tm->ftm_fab->fab_fi = fi;
+	fi_freeinfo(hints);
+	
+	rc = fi_fabric(tm->ftm_fab->fab_fi->fabric_attr, &tm->ftm_fab->fab_fab,
+		       NULL);
+	if (rc != FI_SUCCESS) {
+		libfab_pep_param_free(pep, tm);
+		return M0_RC(rc);
+	}
+
+	rc = libfab_waitfd_init(tm);
+	if (rc != FI_SUCCESS) {
+		libfab_pep_param_free(pep, tm);
+		return M0_RC(rc);
+	}
+
+	rc = fi_passive_ep(tm->ftm_fab->fab_fab, tm->ftm_fab->fab_fi,
+			   &pep->pep_pep, NULL);
+	if (rc != FI_SUCCESS) {
+		libfab_pep_param_free(pep, tm);
+		return M0_RC(rc);
+	}
+
+	rc = libfab_pep_res_init(pep, tm);
+	if (rc != FI_SUCCESS) {
+		libfab_pep_param_free(pep, tm);
+		return M0_RC(rc);
+	}
+
+	rc = fi_listen(pep->pep_pep);
+	if (rc != FI_SUCCESS) {
+		libfab_pep_param_free(pep, tm);
+		return M0_RC(rc);
+	}
+
+	rc = fi_domain(tm->ftm_fab->fab_fab, tm->ftm_fab->fab_fi,
+		       &tm->ftm_fab->fab_dom, NULL);
+	if (rc != FI_SUCCESS) {
+		M0_LOG(M0_ERROR," \n fi_domain = %d \n ", rc);
+		libfab_pep_param_free(pep, tm);
+		return M0_RC(rc);
+	}
+
+	rc = fi_srx_context(tm->ftm_fab->fab_dom, tm->ftm_fab->fab_fi->rx_attr,
+			    &tm->ftm_rctx, NULL);
+	if (rc != FI_SUCCESS) {
+		M0_LOG(M0_ERROR," \n fi_srx_context = %d \n ", rc);
+		libfab_pep_param_free(pep, tm);
+		return M0_RC(rc);
+	}
+
+	rc = libfab_tm_res_init(tm);
+	fab_sndbuf_tlist_init(&ep->fep_sndbuf);
+
```

Comment:
if rc is error,
libfab_pep_param_free() should be called to do cleanup.

Otherwise, ep->fep_listen and ep->fep_listen->pep_aep are leaked.

Reply:
Corrected